src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java

```diff
@@ -176,118 +173,107 @@ public void write(JsonWriter out, Optional<A> optional) throws IOException {
-                JsonObject stats = jsonElement.getAsJsonObject();
-                Applications app = gson.fromJson(stats.get(CP_APPLICATIONS),
-                        Applications.class);
-                HttpServer server = null;
-                for (Map.Entry<String, JsonElement> entry : stats.entrySet()) {
-                    String key = entry.getKey();
-                    if (key.startsWith(CP_SERVER_PREFIX)) {
-                        server = gson.fromJson(entry.getValue(), HttpServer.class);
-                        break;
-                    }
+        public void write(JsonWriter jsonWriter, Stats stats) throws IOException {
+            throw new UnsupportedOperationException("Writing JSON not supported.");
+        }
+
+        @Override
+        public Stats read(JsonReader jsonReader) throws IOException {
+            Applications app = null;
+            HttpServer server = null;
+            jsonReader.beginObject();
+            while (jsonReader.hasNext()) {
+                String name = jsonReader.nextName();
+                if (name.equals(CP_APPLICATIONS)) {
+                    app = gson.fromJson(jsonReader, Applications.class);
+                } else if (name.startsWith(CP_SERVER_PREFIX)) {
+                    server = gson.fromJson(jsonReader, HttpServer.class);
+                } else {
+                    jsonReader.skipValue();
                 }
-                return new Stats(app, server);
-            } catch (Exception e) {
-                throw new JsonParseException(e);
             }
+            jsonReader.endObject();
+            return new Stats(app, server);
         }
     }
 
     /**
-     * Deserializer for Arguments class.
+     * Json TypeAdapter for Arguments class.
      * Breaks the incoming arguments into args and kwargs parts
      * and fills a new Arguments instance.
      */
-    private class ArgumentsDeserializer implements JsonDeserializer<Arguments> {
+    private class ArgumentsAdapter extends TypeAdapter<Arguments> {
 
         private static final String KWARG_KEY = "__kwarg__";
 
         @Override
-        public Arguments deserialize(JsonElement json, Type typeOfT,
-                JsonDeserializationContext context) throws JsonParseException {
-            Arguments result = new Arguments();
+        public void write(JsonWriter jsonWriter, Arguments args) throws IOException {
+            throw new UnsupportedOperationException("Writing JSON not supported.");
+        }
 
-            if (json != null && json.isJsonArray()) {
-                for (JsonElement jsonElement : json.getAsJsonArray()) {
-                    fillArgs(result, jsonElement);
+        @Override
+        public Arguments read(JsonReader jsonReader) throws IOException {
+            if (jsonReader.peek() == JsonToken.NULL) {
+                throw new JsonParseException("null is not a valid value for Arguments");
+            } else if (jsonReader.peek() != JsonToken.BEGIN_ARRAY) {
```

Comment:
I reconsidered this check, I think to maintain the old behaviour we should consume whatever is next in the stream and then return null. Is this the desired behaviour though? Whether we keep it as it is now, throw an exception like with the null check, or just continue without the check all together, the outcome will likely be the same; parse error. Unless you strongly want to keep it as it was, I'd say I might as well just remove this check and let beingArray() fail.